File: app/src/main/java/com/gmail/fattazzo/formula1world/fragments/stats/drivers/DriversNationalityWinsStatsFragment.java
package com.gmail.fattazzo.formula1world.fragments.stats.drivers;

import android.graphics.Bitmap;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.widget.BaseAdapter;

import com.github.mikephil.charting.formatter.DefaultValueFormatter;
import com.github.mikephil.charting.formatter.IValueFormatter;
import com.gmail.fattazzo.formula1world.R;
import com.gmail.fattazzo.formula1world.ergast.imagedb.service.stats.StatsData;
import com.gmail.fattazzo.formula1world.fragments.stats.AbstractStatsChartFragment;
import com.gmail.fattazzo.formula1world.fragments.stats.adapters.StatsDataImageLabelListAdapter;
import com.gmail.fattazzo.formula1world.utils.CountryNationality;
import com.gmail.fattazzo.formula1world.utils.ImageUtils;
import com.gmail.fattazzo.formula1world.utils.Utils;

import org.androidannotations.annotations.Bean;
import org.androidannotations.annotations.EFragment;

import java.text.DecimalFormat;
import java.util.List;

/**
 * @author fattazzo
 *         <p/>
 *         date: 24/08/17
 */
@EFragment(R.layout.fragment_stats_chart)
public class DriversNationalityWinsStatsFragment extends AbstractStatsChartFragment {

    @Bean
    protected ImageUtils imageUtils;

    @Bean
    protected Utils utils;

    @NonNull
    @Override
    public List<StatsData> loadData() {
        return statisticsService.loadDriversWinsNationality(seasonStart, seasonEnd);
    }

    @NonNull
    @Override
    protected IValueFormatter getChartValueFormatter() {
        return new DefaultValueFormatter(0);
    }

    @NonNull
    @Override
    protected DecimalFormat getListValueFormat() {
        return new DecimalFormat("0");
    }

    @Override
    protected BaseAdapter createListAdapter(List<StatsData> data, DecimalFormat valueFormat) {
        return new StatsDataImageLabelListAdapter(getActivity(), data, valueFormat) {
            @Nullable
            @Override
            public Bitmap getImage(StatsData data) {
                CountryNationality country = utils.getCountryNationality(data.getLabel());
                if (country != null) {
                    return imageUtils.getFlagForCountryCode(country.getAlpha2Code());
                }
                return null;
            }
        };
    }
}

File: app/src/main/kotlin/com/gmail/fattazzo/formula1world/fragments/stats/drivers/DriversNationalityWinsStatsFragment.kt
/*
 * Project: total-gp-world
 * File: DriversNationalityWinsStatsFragment.kt
 *
 * Created by fattazzo
 * Copyright © 2018 Gianluca Fattarsi. All rights reserved.
 *
 * MIT License
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.gmail.fattazzo.formula1world.fragments.stats.drivers

import android.graphics.Bitmap
import android.widget.BaseAdapter
import com.github.mikephil.charting.formatter.DefaultValueFormatter
import com.github.mikephil.charting.formatter.IValueFormatter
import com.gmail.fattazzo.formula1world.R
import com.gmail.fattazzo.formula1world.ergast.imagedb.service.stats.StatsData
import com.gmail.fattazzo.formula1world.fragments.stats.AbstractStatsChartFragment
import com.gmail.fattazzo.formula1world.fragments.stats.adapters.StatsDataImageLabelListAdapter
import com.gmail.fattazzo.formula1world.utils.ImageUtils
import com.gmail.fattazzo.formula1world.utils.Utils
import org.androidannotations.annotations.Bean
import org.androidannotations.annotations.EFragment
import java.text.DecimalFormat

/**
 * @author fattazzo
 *
 *
 * date: 24/08/17
 */
@EFragment(R.layout.fragment_stats_chart)
open class DriversNationalityWinsStatsFragment : AbstractStatsChartFragment() {
    override val chartValueFormatter: IValueFormatter
        get() = DefaultValueFormatter(0)
    override val listValueFormat: DecimalFormat
        get() = DecimalFormat("0")

    @Bean
    lateinit protected var imageUtils: ImageUtils

    @Bean
    lateinit protected var utils: Utils

    public override fun loadData(): List<StatsData> {
        return statisticsService.loadDriversWinsNationality(this.seasonStart!!, this.seasonEnd!!)
    }

    override fun createListAdapter(data: List<StatsData>, valueFormat: DecimalFormat): BaseAdapter {
        return object : StatsDataImageLabelListAdapter(activity!!, data, valueFormat) {
            override fun getImage(data: StatsData): Bitmap? {
                val country = utils.getCountryNationality(data.label)
                return if (country != null) {
                    imageUtils.getFlagForCountryCode(country.alpha2Code!!)
                } else null
            }
        }
    }
}
